consent/strategy_default.go

```diff
@@ -1178,3 +1214,136 @@ func (s *DefaultStrategy) loginSessionFromCookie(r *http.Request) *flow.LoginSes
 
 	return ls
 }
+
+func (s *DefaultStrategy) requestDevice(ctx context.Context, w http.ResponseWriter, r *http.Request, req fosite.Requester) error {
+	return s.forwardDeviceRequest(ctx, w, r, req)
+}
+
+func (s *DefaultStrategy) forwardDeviceRequest(ctx context.Context, w http.ResponseWriter, r *http.Request, req fosite.Requester) error {
+	// Set up csrf/challenge/verifier values
+	verifier := strings.Replace(uuid.New(), "-", "", -1)
+	challenge := strings.Replace(uuid.New(), "-", "", -1)
+	csrf := strings.Replace(uuid.New(), "-", "", -1)
+
+	// Generate the request URL
+	iu := s.c.OAuth2DeviceAuthorisationURL(ctx)
+	iu.RawQuery = r.URL.RawQuery
+
+	if err := s.r.ConsentManager().CreateDeviceGrantRequest(
+		r.Context(),
+		&flow.DeviceGrantRequest{
+			ID:         challenge,
+			Verifier:   verifier,
+			CSRF:       csrf,
+			RequestURL: iu.String(),
+		},
+	); err != nil {
+		return errorsx.WithStack(err)
+	}
+
+	store, err := s.r.CookieStore(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := createCsrfSession(w, r, s.r.Config(), store, s.r.Config().CookieNameDeviceVerifyCSRF(ctx), csrf, s.c.ConsentRequestMaxAge(ctx)); err != nil {
+		return errorsx.WithStack(err)
+	}
+
+	query := url.Values{"device_challenge": {challenge}}
+	if r.URL.Query().Has("user_code") {
+		query.Add("user_code", r.URL.Query().Get("user_code"))
+	}
+
+	http.Redirect(w, r, urlx.SetQuery(s.c.DeviceUrl(ctx), query).String(), http.StatusFound)
```

Comment:
`DeviceUrl` should be the URL that shows a UI to enter the user code, right? Maybe name it `UserCodeURL`?

Reply:
Makes sense now, please ignore :)